theseus/labs/lie/functional/so3_impl.py

```diff
@@ -327,14 +323,14 @@ def backward(cls, ctx, grad_output):
 # -----------------------------------------------------------------------------
 # Logarithm Map
 # -----------------------------------------------------------------------------
-def _log_impl(group: torch.Tensor) -> torch.Tensor:
-    check_group_tensor(group)
-    sine_axis = group.new_zeros(group.shape[0], 3)
-    sine_axis[:, 0] = 0.5 * (group[:, 2, 1] - group[:, 1, 2])
-    sine_axis[:, 1] = 0.5 * (group[:, 0, 2] - group[:, 2, 0])
-    sine_axis[:, 2] = 0.5 * (group[:, 1, 0] - group[:, 0, 1])
-    cosine = 0.5 * (group[:, 0, 0] + group[:, 1, 1] + group[:, 2, 2] - 1)
-    sine = sine_axis.norm(dim=1)
+def _log_impl_helper(group: torch.Tensor):
+    size = group.shape[:-2]
+    sine_axis = group.new_zeros(size + (3,))
+    sine_axis[..., 0] = 0.5 * (group[..., 2, 1] - group[..., 1, 2])
+    sine_axis[..., 1] = 0.5 * (group[..., 0, 2] - group[..., 2, 0])
+    sine_axis[..., 2] = 0.5 * (group[..., 1, 0] - group[..., 0, 1])
+    cosine = 0.5 * (group[..., 0, 0] + group[..., 1, 1] + group[..., 2, 2] - 1)
```

Comment:
Just noticed this for the first time. Wondering it the code below will be more efficient because the indexing is happening under the hood by torch.

```python
cosine = 0.5 * (group.diagonal(dim1=-1, dim2=-2).sum(dim=-1) - 1)
```

Reply:
Good catch. Just refactored the code.